components/cam/src/dynamics/se/fv_physics_coupling_mod.F90

```diff
@@ -0,0 +1,284 @@
+! require special consideration of the model's vertical coordinate, which is 
+! currently based on moist pressure. In either case, dyn_to_fv_phys() would stay
+! the same, with simple weighted averaging over element subcells.
+! 
+! Author: Walter Hannah (LLNL)
+!---------------------------------------------------------------------------------------------------
+module fv_physics_coupling_mod
+  use element_mod,    only: element_t
+  use shr_kind_mod,   only: r8=>shr_kind_r8
+  use kinds,          only: real_kind, int_kind
+  use constituents,   only: pcnst, cnst_name
+  use dimensions_mod, only: np, npsq, nelemd, nlev
+  use dyn_grid,       only: fv_nphys
+  use ppgrid,         only: pcols, pver, pverp
+  
+  private
+
+  ! These method encapsulate the coupling method for the fv phys grid,
+  ! they are used by d_p_coupling() and p_d_coupling() in dp_coupling.F90 
+  public :: fv_phys_to_dyn
+  public :: fv_phys_to_dyn_topo
+  public :: dyn_to_fv_phys
+
+contains
+  !=================================================================================================
+  !=================================================================================================
+  subroutine fv_phys_to_dyn(elem,T_tmp,uv_tmp,q_tmp)
+    ! Purpose: Copy physics state to dynamics grid
+    use control_mod,    only: ftype
+    use dyn_comp,       only: TimeLevel
+    use derivative_mod, only: subcell_integration
+    implicit none
+    !---------------------------------------------------------------------------
+    ! interface arguments
+    type(element_t),      intent(inout) :: elem(:)          ! dynamics element structure
+    real(kind=real_kind), intent(inout) :: T_tmp (:,:,:)    ! temp array to hold T
+    real(kind=real_kind), intent(inout) :: uv_tmp(:,:,:,:)  ! temp array to hold u and v
+    real(kind=real_kind), intent(inout) :: q_tmp (:,:,:,:)  ! temp to hold advected constituents
+    ! local variables
+    integer(kind=int_kind) :: ie, m, i, j, icol, ilyr       ! loop iterators
+    integer(kind=int_kind) :: ii, jj, gi, gj                ! GLL loop iterator and indices for pg2
+    integer                :: tl_f
+    real(kind=real_kind), dimension(fv_nphys*fv_nphys,pver,pcnst) :: qo_phys ! reconstructed initial physics state 
+    real(r8), dimension(np,np)             :: dp_gll
+    real(r8), dimension(fv_nphys,fv_nphys) :: inv_dp_fvm
+    !---------------------------------------------------------------------------
+    ! Copy tendencies on the physics grid over to the dynamics grid (GLL)
+    !---------------------------------------------------------------------------
+    tl_f = TimeLevel%n0
+    do ie = 1,nelemd
+      !-------------------------------------------------------------------------
+      ! Recalculate state that was previously sent to physics 
+      !-------------------------------------------------------------------------
+      if (ftype==2.or.ftype==3) then
```

Comment:
Could you please explain why this code depends on ftype? Naively, I assume this is remapping code, mapping from D to P may depend on ftype, but this is P to D? Unrelated, but ftype=3 is not used anymore.

Reply:
Because of tracers, I guess.

Reply:
Hopefully I understand this correctly... but with ftype=1 the dycor only needs the tracer tendency, and thus there's no need for this correction. But with ftype=2 the dycor does it's own calculation of the tendency from the new tracer state.

Without this the correction the new tracer state is not consistent with the "old" tracer state on the dynamics grid. Therefore the "old" tracer state on the physics grid needs to be calculated (i.e. qo) and replaced with the "old" tracer state that currently resides on the dynamics grid.

This correction was actually inspired by studying your branch. It fixed a puzzling behavior in the pg2 runs in which precipitation was strongly under-estimated, although still generally reasonable.

Reply:
I should probably put a better explanation in the comments.

Reply:
We probably should document this somewhere in a long run. Issues come from tendency/state in cam output and different ftypes, but it is hard to keep track of this. Would you please replace ftype3 with ftype4 here? ftype3 is not used anymore, but ftype4 is going to be. There is also ftype0, I will try to compose some notes on all this...

Reply:
ok, I'll add ftype=4, but shouldn't we keep ftype=3 just in case someone tries to test it?

Could I just use the condition of ftype>=2?

Reply:
ftype3 should have been erased from everywhere, if I did not get rid of it in cam configs, then I should have. So, no one should try to run with ftype3. Also, we do not have a guideline that ftype>=2 is 'cam state' logic, but >= sign is up to you.

Reply:
ohhhh, I see. I'll take your original suggestion then.